### PR TITLE
add word hashing processor unit

### PR DESCRIPTION
**Pull request by bwanglzu**
Add `WordHashing` unit for DSSM and CDSSM.

For issue #218 

Also #245

### CODE REVIEW

File: matchzoo/preprocessor/process_units.py [hunk 1]
```diff
@@ -295,3 +297,62 @@ def fit(self, tokens: list):
...
+        Transform list of tri-letters into word hashing layer.
+
+        :param tri_letters: list of `tri_letters` generated by
+        :class:`NgramLetterUnit`.
+        :return: Word hashing representation of `tri-letters`.
+        """
+        counted_tri_letters = dict(collections.Counter(tri_letters))
+        for key, value in counted_tri_letters.items():
+            # get index.
+            idx = self._term_index[key]
```
[the remark this answers is not in the document]
Comment (bwanglzu) [reply matchzoo/preprocessor/process_units.py]: `VocabularyUnit` is an essential prerequisite for `WordHashingUnit`.

File: matchzoo/preprocessor/process_units.py [hunk 2]
```diff
@@ -295,3 +297,62 @@ def fit(self, tokens: list):
...
+       >>> hashing[0]
+       0.0
+       >>> hashing[1]
+       1.0
+       >>> hashing.shape
+       (5,)
+
+    """
+
+    def __init__(
```
[the remark this answers is not in the document]
Comment (bwanglzu) [reply matchzoo/preprocessor/process_units.py]: I'll adjust this part in another branch & pull request.

File: matchzoo/preprocessor/process_units.py [hunk 3]
```diff
@@ -295,3 +297,66 @@ def fit(self, tokens: list):
...
+    ) -> np.ndarray:
+        """
+        Transform list of :attr:`tri-letters` into word hashing layer.
+
+        :param tri_letters: list of `tri_letters` generated by
+        :class:`NgramLetterUnit`.
+        :return: Word hashing representation of `tri-letters`.
+        """
+        hashing = np.zeros(len(self._term_index))
+        counted_tri_letters = dict(collections.Counter(tri_letters))
```
Comment (uduse) [inline matchzoo/preprocessor/process_units.py]: `Counter` is a `dict` subclass.

File: matchzoo/preprocessor/process_units.py [hunk 4]
```diff
@@ -295,3 +297,66 @@ def fit(self, tokens: list):
...
+        for key, value in counted_tri_letters.items():
+            # get index.
+            idx = self._term_index[key]
+            # update hashing layer.
```
Comment (uduse) [inline matchzoo/preprocessor/process_units.py]: IMHO, these are redundant comments.

**Pull request merged**